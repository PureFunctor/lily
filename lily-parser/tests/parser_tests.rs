use lily_parser::parse_top_level;

#[test]
pub fn top_level_0() {
    let source = r"
infixl 4 add as +

example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_1() {
    let source = r"
example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_2() {
    let source = r"
infixl 1 add as +
infixl 2 add as *

example = a + b * (c + d) + e
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_3() {
    let source = r"
infixl 1 add as +

example = if if a then b else c then if d then e else f else if g then h else i
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_4() {
    let source = r"
example = f a b
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_5() {
    let source = r"
example = f if a then b else c d

example = f (if a then b else c) d
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}
