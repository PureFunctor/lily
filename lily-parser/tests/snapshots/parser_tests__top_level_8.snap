---
source: lily-parser/tests/parser_tests.rs
expression: parse_top_level(source)
---
Ok(
    Module {
        declarations: [
            Declaration {
                begin: 1,
                end: 89,
                kind: ValueDeclaration(
                    "example",
                    [],
                    Expression {
                        begin: 11,
                        end: 89,
                        kind: CaseOf(
                            [
                                Expression {
                                    begin: 16,
                                    end: 17,
                                    kind: Variable(
                                        "a",
                                    ),
                                },
                                Expression {
                                    begin: 19,
                                    end: 20,
                                    kind: Variable(
                                        "b",
                                    ),
                                },
                            ],
                            [
                                CaseArm {
                                    patterns: [
                                        GreaterPattern {
                                            begin: 26,
                                            end: 34,
                                            kind: Application(
                                                GreaterPattern {
                                                    begin: 26,
                                                    end: 30,
                                                    kind: Constructor(
                                                        "Cons",
                                                    ),
                                                },
                                                [
                                                    GreaterPattern {
                                                        begin: 31,
                                                        end: 32,
                                                        kind: Variable(
                                                            "a",
                                                        ),
                                                    },
                                                    GreaterPattern {
                                                        begin: 33,
                                                        end: 34,
                                                        kind: Variable(
                                                            "b",
                                                        ),
                                                    },
                                                ],
                                            ),
                                        },
                                        GreaterPattern {
                                            begin: 36,
                                            end: 37,
                                            kind: Null,
                                        },
                                    ],
                                    condition: Some(
                                        Expression {
                                            begin: 41,
                                            end: 46,
                                            kind: Variable(
                                                "hello",
                                            ),
                                        },
                                    ),
                                    expression: Expression {
                                        begin: 50,
                                        end: 62,
                                        kind: DoBlock(
                                            [
                                                DoStatement {
                                                    begin: 57,
                                                    end: 62,
                                                    kind: DiscardExpression(
                                                        Expression {
                                                            begin: 57,
                                                            end: 62,
                                                            kind: Variable(
                                                                "world",
                                                            ),
                                                        },
                                                    ),
                                                },
                                            ],
                                        ),
                                    },
                                },
                                CaseArm {
                                    patterns: [
                                        GreaterPattern {
                                            begin: 65,
                                            end: 66,
                                            kind: Null,
                                        },
                                        GreaterPattern {
                                            begin: 68,
                                            end: 71,
                                            kind: Constructor(
                                                "Nil",
                                            ),
                                        },
                                    ],
                                    condition: None,
                                    expression: Expression {
                                        begin: 75,
                                        end: 89,
                                        kind: DoBlock(
                                            [
                                                DoStatement {
                                                    begin: 82,
                                                    end: 89,
                                                    kind: DiscardExpression(
                                                        Expression {
                                                            begin: 82,
                                                            end: 89,
                                                            kind: Variable(
                                                                "example",
                                                            ),
                                                        },
                                                    ),
                                                },
                                            ],
                                        ),
                                    },
                                },
                            ],
                        ),
                    },
                ),
            },
        ],
    },
)
